Reusing rust type binding

anforowicz opened the pull request:
@zetafunction, can you PTAL and do an overall review from your side before I submit this as an upstream PR?  

We are still trying to figure out the process of contributing to `cxx`:

* I think it makes sense (as we suggested in the email) to do an internal review before opening upstream PRs.  This way we save some of dtolnay's time on things that we could have just caught/tweaked ourselves.
* I am not sure exactly how to do such internal reviews.  I think it's fine to raise a PR against my master branch (like _this_ PR), and then throw away such a PR after addressing the review feedback.

WDYT?

anforowicz committed:
Rename `aliases` field in `Types` struct to `cxx_aliases`.

anforowicz committed:
Support for type aliases in `extern "Rust"` section.

Fixes https://crbug.com/433254489

anforowicz committed:
Add minimal test for parsing of type aliases in `extern "C++"` section.

anforowicz committed:
Deduplicate destructuring `syntax::file::Module` into `parse_items`.

`syntax::parse::parse_items` is currently called from two places:
from `macro/src/expand.rs` and from `gen/src/mod.rs`.  (In the future
it may be also called from a `syntax/test_support.rs` helper.)

Before this commit, all those places had to destructure/interpret
a `syntax::file::Module` into `content`, `namespace`, and `trusted`.

After this commit, this destructuring/interpretation is deduplicated
into `syntax::parse::parse_items`.  This requires some minor gymnastics:

* `gen/src/mod.rs` has to call `std::mem::take(&mut bridge.attrs)`
  instead of doing a partial destructuring and passing `bridge.attrs`
  directly.  This is an extra complexity, but we already do the same
  thing in `macro/src/expand.rs` before this commit, so hopefully this
  is ok.
* `fn parse_items` takes `&mut Module` rather than `Module` and
  "consumes" / `std::mem::take`s `module.content`, because
  `macro/src/expand.rs` needs to retain ownership of other `Module`
  fields.  This also seems like an unfortunate extra complexity, but
  (again) before this commit we already do this for `bridge.attrs` in
  `macro/src/expand.rs`.

zetafunction commented on book/src/extern-rust.md:
```diff
@@ -163,3 +163,46 @@ mod ffi {
+
+Extern Rust types support a syntax for declaring that a C++ binding of the
+given Rust type already exists outside of the current bridge module. This
+avoids generating a fresh new binding which would clash with the first
```
Minor wording suggestion: s/fresh new binding/separate binding/

zetafunction commented on syntax/parse.rs:
```diff
@@ -27,12 +27,11 @@ pub(crate) mod kw {
-    trusted: bool,
-    namespace: &Namespace,
-) -> Vec<Api> {
+pub(crate) fn parse_items(cx: &mut Errors, bridge: &mut Module) -> Vec<Api> {
```
... I think I'm fine with this, but I don't love that we're passing a `Module` here but just ignoring `attrs`.

... is there a way to logically group the three fields needed by this into a struct and then just pass the struct here? Then we don't have to use `std::mem::take()` at all, I think.

anforowicz replied:
Thank you for the feedback.  I think this is a valid feedback.  But I struggle to act on it and I think I'll just submit the PR as-is.

# Extracting sub-structs

Extracting/grouping sub-structs is tricky.  We could in theory have 3 sub-structs: attrs, namespace+unsafety+contents, all-other-tokens.  But this split is a bit arbitrary (`unsafe` token influences parsing of APIs, but other than that why would we split it from all the other tokens;  and namespace is kind of sort-of parsed from attrs, except a bit differently in proc-macro case vs cmdbridge case).

# Namespace parsing in tests

When reading the code I've discovered that `syntax/test_support.rs` doesn't parse the namespace in `#[cxx::bridge(namespace = "foo")]`.  The fact that namespace is not parsed by `syn`/`parse` of `syntax::file::Module` caught me by surprise, but it seems to make sense.

I've added an extra commit to add namespace parsing to `test_support.rs`.

# Passing `&Module` by reference

If we want to avoid this icky business of leaving `Module` in a partially empty/taken/restructured state, then I think we could consider making `attrs::parse` and `syntax::parse_items` take their arguments by reference (rather than by value).  Maybe this is something to pursue as a follow-up?  This would have cascading effects (probably requiring `.clone()` deeper in `syntax/parse.rs` and `syntax/attrs.rs`) so I don't plan to proactively try tackling this.

zetafunction replied:
I don't love the partial consumption. If it were up to me, I'd probably just copy and paste the manual destructuring into more places if we can't think of a logical grouping. But I'm not strongly against this either.

zetafunction commented on syntax/test_support.rs:
```diff
@@ -0,0 +1,33 @@
+/// Parses a `TokenStream` containing `#[cxx::bridge] mod { ... }`.
+pub fn parse_apis(cxx_bridge: TokenStream) -> syn::Result<Vec<Api>> {
+    let mut module = syn::parse2::<Module>(cxx_bridge)?;
+
```
Very minor but it seems like this newline could be dropped (I see line 13 and line 15 as part of the same logical block for setting up for line 16)

zetafunction commented on gen/src/mod.rs:
There's a small typo in the commit message (gymnatics -> gymnastics)

(GH doesn't seem to support commenting on commit messages)

zetafunction commented on macro/src/expand.rs:
```diff
@@ -1275,23 +1275,34 @@ fn expand_type_alias_verify(alias: &TypeAlias, types: &Types) -> TokenStream {
     let type_id = type_id(&alias.name);
```
I think this can be moved to a narrowed scope, inside the `Lang::Cxx | Lang::CxxUnwind`

anforowicz replied:
Good catch - thanks!

zetafunction commented on tests/ffi/lib.rs:
```diff
@@ -342,6 +342,13 @@ pub mod ffi {
     impl CxxVector<SharedString> {}
+
+    extern "Rust" {
+        type RustTypeBindingReusedAcrossModules = crate::module::RustTypeBindingReusedAcrossModules;
```
This feels a bit wordy to me; I wonder if something like `CrossModuleRustType` would work (admittedly it's not much shorter though)

zetafunction commented on tests/ffi/lib.rs:
```diff
@@ -342,6 +342,13 @@ pub mod ffi {
+
+    extern "Rust" {
+        type RustTypeBindingReusedAcrossModules = crate::module::RustTypeBindingReusedAcrossModules;
+        fn r_get_value_from_rust_type_binding_reused_across_modules(
```
Similar thought here: `r_get_value_from_cross_module_rust_type()`

zetafunction commented on tests/ffi/module.rs:
```diff
@@ -78,3 +78,24 @@ pub mod ffi2 {
+        type RustTypeBindingReusedAcrossModules;
+
+        #[allow(clippy::unnecessary_box_returns)]
+        fn r_new_rust_type_binding_reused_across_modules(
```
And `r_boxed_cross_module_rust_type()`

anforowicz committed:
Fix how `test_support.rs` parses namespace from `#[cxx::bridge]`.

Before this commit `fn parse_apis` in `syntax/test_support.rs` would
ignore the namespace in `#[cxx::bridge(namespace = "ignored")]`.
In other words, the new test added in `syntax/namespace.rs` would
fail before this commit.  This commit fixes this.

At a high-level this commit:

* Moves two pieces of code from `gen/src/file.rs` into `syntax/...`
  to make them reusable from `syntax/test_support.rs`:
    - `fn find_cxx_bridge_attr` (moved into `syntax/attrs.rs`)
    - `Namespace::parse_attr` (moved into
      `syntax/namespace.rs`)
* Reuses these pieces of code from `syntax/test_support.rs`
* Renames `Namespace::parse_bridge_attr_namespace` to
  `Namespace::parse_stream` so that all 3 parse methods are named
  after their input: `parse_attr`, `parse_stream`, `parse_meta`.
* Adds a `syntax/`-level unit test that verifies that the namespace
  is indeed getting correctly parsed and propagated

anforowicz commented:
@zetafunction, can you PTAL at the new commit at https://github.com/anforowicz/cxx/pull/2/commits/4288f636f14f8eae548d424e83dc8a03b5b4379d?

Other than that, I think I addressed most of your feedback (other than the one about partially-taking/destructuring `Module` fields... :-/ - see my comment above with an explanation).

zetafunction commented on gen/src/file.rs:
Comment on the commit description:

> Fix how test_support.rs parses namespace from #[cxx::bridge].

It's not entirely clear what this is fixing to me; it just seems like refactoring.

anforowicz replied:
> Comment on the commit description:
> 
> > Fix how test_support.rs parses namespace from #[cxx::bridge].
> 
> It's not entirely clear what this is fixing to me

I tried to improve the commit message to better explain what was broken before this commit.

> it just seems like refactoring.

The fix is the presence of `module.namespace = ...` that got added to `syntax/test_support.rs`.

anforowicz commented:
FWIW I've also wondered whether we should also keep the convention from [`enum Api`](https://github.com/dtolnay/cxx/blob/68b827689c684e76d7d28374a006c0441380e9ae/syntax/mod.rs#L52-L63) which has variants like 1) `RustFunction` and `CxxFunction`, or 2) `RustType` and `CxxType`, but 3) only a single `TypeAlias` variant (which after this PR has a `lang` field).  I think I'll try putting together a commit that bifurcates the `TypeAlias` variant into `CxxTypeAlias` and `RustTypeAlias`, but I am not yet sure if this should be part of the final upstream PR.

anforowicz commented:
Thank you for the review feedback.  I've now opened an upstream PR at https://github.com/dtolnay/cxx/pull/1539.  So let me close _this_ PR.
